SEED-platform/seed
# OGBS - French (Canada) localization

#### Any background context you want to provide?

To be acceptable for use in Canada, the user interface must be available in both official languages: French and English.

#### What's this PR do?

This PR extends seed to use browser `Accept:` headers to deliver `fr-CA` (or, effectively as a no-op, `en-US`) language localization. User interface strings are being managed at [lokalise.co](https://lokalise.co/project/3537487659ca9b1dce98a7.36378626/?view=multi) on their free enterprise account for open source projects; please contact @rgm for access.

(Date/currency/number localization (eg `1 000,00$` or `12 Mai 2017`) is not addressed here and will be taken up in another PR).

Two significant outstanding issues for discussion, semi-related:
1. Minor XSS vector through the `sce` sanitization strategy
2. Handling column names ... just leaving untranslated for now and will handle as part of sorting out the remainder of the Pint column name work.

#### How should this be manually tested?

1. Flip your browser's preference to put `fr` or `fr_CA` as the top preference.
2. Reload SEED.
3. Browse the site, trying out normal workflows.
4. Flip browser's preference back to `en_US` or just entirely reset browser.

Expected behaviour:

All UI strings should appear in French or English as chosen. There should be no English in the French run-through; this would indicate a missing string in the localization file as the UI will fall back to English if the French string is missing. There should be no strings of the sort `ALL_CAPS_SHORT_DESCRIPTION_OF_CONTENT` in either English or French; this would indicate that the translation has failed entirely.

#### What are the relevant tickets?

n/a.

#### Screenshots (if appropriate)

<img width="1550" alt="screen shot 2017-11-06 at 23 07 18" src="https://user-images.githubusercontent.com/6805/32481244-71c03342-c347-11e7-8d5e-96ccecf8af35.png">
<img width="1550" alt="screen shot 2017-11-06 at 23 07 55" src="https://user-images.githubusercontent.com/6805/32481254-7b3c4514-c347-11e7-8eeb-2f7a7c7271fd.png">
<img width="1550" alt="screen shot 2017-11-06 at 23 08 24" src="https://user-images.githubusercontent.com/6805/32481262-84cdfff0-c347-11e7-8461-d6b7458d18ac.png">
<img width="1550" alt="screen shot 2017-11-06 at 23 06 44" src="https://user-images.githubusercontent.com/6805/32481273-8d3d60a4-c347-11e7-8764-3104d3f4099a.png">

#### Definition of Done:

Yes, package.json and bower.json both contain the new dependencies. Nothing new has been added in the Django layer.
